core/src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregator.java

```diff
@@ -183,24 +211,571 @@ protected void doClose() {
+                FieldDataFieldsContext fieldDataFieldsContext = subSearchContext
+                        .getFetchSubPhaseContext(FieldDataFieldsFetchSubPhase.CONTEXT_FACTORY);
+                for (String field : fieldDataFields) {
+                    fieldDataFieldsContext.add(new FieldDataField(field));
+                }
+                fieldDataFieldsContext.setHitExecutionNeeded(true);
+            }
+            if (scriptFields != null) {
+                for (ScriptField field : scriptFields) {
+                    SearchScript searchScript = subSearchContext.scriptService().search(subSearchContext.lookup(), field.script(),
+                            ScriptContext.Standard.SEARCH, Collections.emptyMap());
+                    subSearchContext.scriptFields().add(new org.elasticsearch.search.fetch.script.ScriptFieldsContext.ScriptField(
+                            field.fieldName(), searchScript, field.ignoreFailure()));
+                }
+            }
+            if (fetchSourceContext != null) {
+                subSearchContext.fetchSourceContext(fetchSourceContext);
+            }
+            if (highlightBuilder != null) {
+                subSearchContext.highlight(highlightBuilder.build(aggregationContext.searchContext().indexShard().getQueryShardContext()));
+            }
+            return new TopHitsAggregator(aggregationContext.searchContext().fetchPhase(), subSearchContext, name, aggregationContext,
+                    parent, pipelineAggregators, metaData);
         }
 
         @Override
         public AggregatorFactory subFactories(AggregatorFactories subFactories) {
             throw new AggregationInitializationException("Aggregator [" + name + "] of type [" + type + "] cannot accept sub-aggregations");
         }
+
+        @Override
+        protected XContentBuilder internalXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field(SearchSourceBuilder.FROM_FIELD.getPreferredName(), from);
+            builder.field(SearchSourceBuilder.SIZE_FIELD.getPreferredName(), size);
+            builder.field(SearchSourceBuilder.VERSION_FIELD.getPreferredName(), version);
+            builder.field(SearchSourceBuilder.EXPLAIN_FIELD.getPreferredName(), explain);
+            if (fetchSourceContext != null) {
+                builder.field(SearchSourceBuilder._SOURCE_FIELD.getPreferredName(), fetchSourceContext);
+            }
+            if (fieldNames != null) {
+                if (fieldNames.size() == 1) {
+                    builder.field(SearchSourceBuilder.FIELDS_FIELD.getPreferredName(), fieldNames.get(0));
+                } else {
+                    builder.startArray(SearchSourceBuilder.FIELDS_FIELD.getPreferredName());
+                    for (String fieldName : fieldNames) {
+                        builder.value(fieldName);
+                    }
+                    builder.endArray();
+                }
+            }
```

Comment:
This is how it works in the Search request too. I'm ok with writing an array all the time in both places if you would prefer me to make the change?